src/encoded/visualization.py

```diff
@@ -2241,39 +2243,57 @@ def browsers_available(assemblies, status, files, types, item_type=None):
     elif item_type not in VISIBLE_DATASET_TYPES_LC:
             return []
 
-    if not files:
-        return []
-
     browsers = set()
     for assembly in assemblies:
-        mapped_assembly = _ASSEMBLY_MAPPER_FULL[assembly]
+        mapped_assembly = _ASSEMBLY_MAPPER_FULL.get(assembly)
         if not mapped_assembly:
             continue
-        if 'ucsc_assembly' in mapped_assembly:
-            browsers.add('ucsc')
-        if 'ensembl_host' in mapped_assembly:
-            browsers.add('ensembl')
-        if 'quickview' in mapped_assembly:
-            browsers.add('quickview')
-
-    if status not in VISIBLE_DATASET_STATUSES:
-        if status not in QUICKVIEW_STATUSES_BLOCKED:
-            return ["quickview"]
-        return []
+        vis_blob = None
+        if request is not None and accession is not None and status == 'released':
+            vis_blob = get_from_es(request, accession + "_" + assembly)  # use of find_or_make_acc_composite() will recurse!
+        if 'ucsc' not in browsers and 'ucsc_assembly' in mapped_assembly.keys():
+            if vis_blob is not None or files is None:  # If not checking files, be lenient
+                browsers.add('ucsc')
+            elif files is not None:
+                for file in files:
+                    if file.get('assembly','') == assembly and file['status'] == 'released' \
+                        and file.get('file_format','') in ['bigBed', 'bigWig']:
+                        browsers.add('ucsc')
+                        break
 
+        if 'ensembl' not in browsers and 'ensembl_host' in mapped_assembly.keys():
+            if vis_blob is not None or files is None:
+                browsers.add('ensembl')
+            elif files is not None:
+                for file in files:
+                    if file.get('assembly','') == assembly and file['status'] == 'released' \
+                        and file.get('file_format','') in ['bigBed', 'bigWig']:
+                        browsers.add('ensembl')
+                        break
+        if 'quickview' not in browsers and 'quickview' in mapped_assembly.keys():
+            # NOTE: quickview may not have vis_blob as 'in progress' files can also be displayed
+            #       Ideally we would also look at files' statuses and formats.  However, the (calculated)files
+            #       property only contains 'released' files so it doesn't really help for quickview!
+            if vis_blob is not None or (status not in QUICKVIEW_STATUSES_BLOCKED):
+                browsers.add('quickview')
+        if len(browsers) == 3:  # No use continuing
```

Comment:
I'm not a fan of this.  I made an explicit check of the values in my example function.